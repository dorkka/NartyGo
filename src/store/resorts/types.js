export const SET_RESORTS = 'resorts/SET_RESORTS';
export const IS_LOADING = 'resorts/IS_LOADING';
export const SET_ERROR = 'resorts/SET_ERROR';
export const SET_RESORT = 'resorts/SET_RESORT';
export const SET_SELECTED_OPTIONS = 'SET_SELECTED_OPTIONS';
