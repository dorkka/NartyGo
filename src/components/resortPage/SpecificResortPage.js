import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ResortWeatherInfo from './ResortWeatherInfo';
import ResortDetails from './ResortDetails';

class SpecificResortPage extends Component {
  state = {
    resort: {},
    weather: {},
    piste: [],
    isLoading: false,
    error: null,
  };

  componentDidMount() {
    this.fetchResort();
  }

  fetchResort() {
    this.setState({ isLoading: true });
    fetch(`http://localhost:3001/resorts/${this.props.match.params.id}`)
      .then((response) => {
        if (response.ok) {
          return response.json();
        }
        let message = 'Something went wrong ...';
        if (response.status === 404) {
          message = 'Page not found 404';
        }
        throw new Error(message);
      })
      .then((resort) => {
        this.setState({
          resort,
          weather: resort.weather,
          piste: resort.piste,
          isLoading: false,
        });
      })
      .catch(error => this.setState({ error, isLoading: false }));
  }

  render() {
    const {
      resort, weather, piste, isLoading, error,
    } = this.state;
    if (error) { return (error.message); }
    if (isLoading) { return <div>Loading in progress</div>; }

    return (
      <div className="container-fluid App">
        <hr />
        <h4>OÅ›rodek narciarski: {resort.name}</h4>
        <div className="row">
          <ResortDetails city={resort.city} piste={piste.length} />
          <ResortWeatherInfo
            temperature={weather.temperature}
            pressure={weather.pressure}
            clouds={weather.clouds}
          />
        </div>
      </div>
    );
  }
}

SpecificResortPage.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string,
    }),
  }).isRequired,
};

export default SpecificResortPage;
