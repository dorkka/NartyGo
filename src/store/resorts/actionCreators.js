import * as types from './types';
import resourceFetcher from '../../services/resourceFetcher';

export const setResorts = (data, headers, perPage) => {
  const pageCount = Math.ceil(headers.get('x-total-count') / perPage);
  const list = [];
  const byId = {};
  data.forEach(element => {
    list.push(element.id);
    byId[element.id] = element;
  });
  return {
    type: types.SET_RESORTS,
    payload: {
      pageCount,
      list,
      byId,
    },
  };
};

export const setError = (error) => ({
  type: types.SET_ERROR,
  error,
});

export const setIsLoading = () => ({
  type: types.IS_LOADING,
});

export const setSpecificResort = data => {
  const byId = { [data.id]: data };
  return {
    type: types.SET_RESORT,
    payload: {
      byId,
    },
  };
};

export const getResorts = (_page, _limit) => (dispatch) => {
  dispatch(setIsLoading());
  return resourceFetcher('http://localhost:3001/resorts')({ _page, _limit })
    .then(({ data, headers }) => {
      dispatch(setResorts(data, headers, _limit));
    })
    .catch(error => dispatch(setError(error)));
};

export const getSpecificResort = (id) => (dispatch, getState) => {
  if ((getState().resorts.byId[id])) {
    return null;
  }
  dispatch(setIsLoading());
  return resourceFetcher(`http://localhost:3001/resorts/${id}`)()
    .then(({ data }) => {
      dispatch(setSpecificResort(data));
    })
    .catch(error => dispatch(setError(error)));
};

export const setSelectOptions = (data) => {
  const resortsOptions = []
  const cityOptions = []
  let label = ''
  data.forEach((element) => {
    label = `${element.name} - ${element.city}`;
    resortsOptions.push({ value: element.name, label, id: element.id })
    cityOptions.push({ value: element.city, label, id: element.id })
  })
  return {
    type: types.SET_SELECTED_OPTIONS,
    payload: {
      resortsOptions,
      cityOptions,
    }
  }
}

export const getAllResorts = () => (dispatch) => {
  resourceFetcher('http://localhost:3001/resorts')().then(({ data }) => {
    dispatch(setSelectOptions(data))
  });
}
