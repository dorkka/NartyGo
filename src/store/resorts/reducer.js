import * as types from './types';

const initialState = {
  byId: {},
  list: [],
  pageCount: 1,
  isLoading: false,
  error: null,
};

function resortsReducer(state = initialState, action) {
  switch (action.type) {
    case types.SET_RESORTS: {
      return {
        ...state,
        byId: {
          ...state.byId,
          ...action.payload.byId,
        },
        list: action.payload.list,
        pageCount: action.payload.pageCount,
        isLoading: false,
      };
    }
    case types.SET_ERROR: {
      return { ...state, error: action.payload.error };
    }
    case types.IS_LOADING: {
      return { ...state, isLoading: true };
    }
    default: {
      return state;
    }
  }
}
export default resortsReducer;
