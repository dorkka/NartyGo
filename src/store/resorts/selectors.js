export const getCurrentResorts = (state) =>
  state.resorts.list.map((id) => state.resorts.byId[id] || {});
export const getSpecificResort = (state, id) =>
  state.resorts.byId[id] || { weather: {}, piste: [] };
