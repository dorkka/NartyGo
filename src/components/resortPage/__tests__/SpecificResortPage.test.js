import { SpecificResortPage } from '../SpecificResortPage';
import { resortId1 } from '../../../specs/fixtures/resorts';

describe('SpecificResortPage', () => {
  const defaultProps = {
    match: {
      params: {
        id: '1',
      },
    },
    getSpecificResort: () => { },
    resort: resortId1,
    isLoading: false,
    error: null,
  };

  const setup = buildSetup(SpecificResortPage, defaultProps, { lifecycleExperimental: true });

  describe('render component', () => {
    test('should render component', () => {
      const { wrapper } = setup();
      expect(wrapper).toMatchSnapshot();
    });

    test('should render error message', () => {
      const { wrapper } = setup({ error: { message: 'test error' } });
      expect(wrapper).toMatchSnapshot();
    });

    test('should render loading in progres when isLoading is truthy', () => {
      const { wrapper } = setup({ isLoading: true });
      expect(wrapper).toMatchSnapshot();
    });

    test('should render loading in progres when cityId is falsy', () => {
      const { wrapper } = setup({ resort: { cityId: NaN } });
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('componentDidMount', () => {
    test('should call getSpecificResort', () => {
      const { props } = setup({ getSpecificResort: jest.fn() });
      expect(props.getSpecificResort).toHaveBeenCalled();
    });

    test('should call getSpecificResort with appriopriate argument', () => {
      const { props } = setup({ getSpecificResort: jest.fn() });
      expect(props.getSpecificResort).toHaveBeenCalledWith('1');
    });
  });
});
