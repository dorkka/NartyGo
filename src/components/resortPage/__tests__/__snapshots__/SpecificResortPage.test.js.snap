// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpecificResortPage render component should render component 1`] = `
<div
  className="container-fluid App"
>
  <hr />
  <h4>
    Ośrodek narciarski: 
    Czantoria
  </h4>
  <div
    className="row"
  >
    <ResortDetails
      city="Ustroń"
      piste={2}
    />
    <Connect(ResortWeather)
      cityId={7531846}
    />
    <div
      className="col-md-4"
    >
      <withProps(withScriptjs(withGoogleMap(Component)))
        coordinates={
          Object {
            "lat": 49.6833162,
            "lng": 18.8196089,
          }
        }
        defaultCenter={
          Object {
            "lat": 49.6833162,
            "lng": 18.8196089,
          }
        }
        zoom={10}
      />
    </div>
  </div>
</div>
`;

exports[`SpecificResortPage render component should render error message 1`] = `"test error"`;

exports[`SpecificResortPage render component should render loading in progres when cityId is falsy 1`] = `
<div>
  Loading in progress
</div>
`;

exports[`SpecificResortPage render component should render loading in progres when isLoading is truthy 1`] = `
<div>
  Loading in progress
</div>
`;
